### PR TITLE
Integrated IPFS as an additional deployment

### PR DESCRIPTION
**Pull request by Clearwood**
Implementing https://github.com/leapdao/leapdao-website/issues/152

Added an additional deployment to IPFS via the pinning service https://pinata.cloud . After the deployment the content hash is used to update the DNS settings through Route 53.

### CODE REVIEW

**Comment by johannbarbie** [comment]
i found the following errors and warnings by travis:
Screen Shot 2020-04-01 at 9 59 08 AM

...review wip...

File: .travis.yml [hunk 1]
```diff
@@ -6,6 +6,7 @@ node_js:
 before_install:
   - pip install --user awscli
   - export PATH=$PATH:$HOME/.local/bin
+  - npm install -g ipfs-deploy
```
Comment (johannbarbie) [inline .travis.yml:9]: - should `ipfs-deploy` be called somewhere?

File: walk.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+//walk.js in project root
```
Comment (johannbarbie) [inline walk.js:1]: does this only replace links that contain the host name? or also absolute path, like `/blog/x`?

Comment (Clearwood) [reply walk.js]: Problematically it did not resolve absolute links into relative ones properly so I took out that part.

**Comment by troggy** [comment]
I'm confused with `pinata.cloud` thing. Aren't we getting dependent on it? If so, how is better then being dependent on AWS S3 ?

File: unpin.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        .then(function (response) {
+          response.data.rows.forEach(function(item){
+            deletePin(item.ipfs_pin_hash)
+          });
+          //console.log(response.data.rows);
+            //handle response here
+        })
+        .catch(function (error) {
+            //handle error here
+        });
```
Comment (troggy) [inline unpin.js:41]:
```suggestion
```

Suggesting not to swallow possible pinata errors, so that build fails if unpinning fails

File: unpin.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                'pinata_api_key': process.env.IPFS_DEPLOY_PINATA__API_KEY,
+                'pinata_secret_api_key': process.env.IPFS_DEPLOY_PINATA__SECRET_API_KEY
+            }
+        })
+        .then(function (response) {
+          response.data.rows.forEach(function(item){
+            deletePin(item.ipfs_pin_hash)
+          });
+          //console.log(response.data.rows);
+            //handle response here
```
Comment (troggy) [inline unpin.js:37]:
```suggestion
```

File: unpin.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                    'pinata_secret_api_key': process.env.IPFS_DEPLOY_PINATA__SECRET_API_KEY
+                }
+            }
+        ).then(function (response) {
+          //console.log(response);
+            //handle response here
+        })
+        .catch(function (error) {
+            //handle error here
+        });
```
Comment (troggy) [inline unpin.js:20]:
```suggestion
```

File: unpin.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            url,
+            {
+                headers: {
+                    'pinata_api_key': process.env.IPFS_DEPLOY_PINATA__API_KEY,
+                    'pinata_secret_api_key': process.env.IPFS_DEPLOY_PINATA__SECRET_API_KEY
+                }
+            }
+        ).then(function (response) {
+          //console.log(response);
+            //handle response here
```
Comment (troggy) [inline unpin.js:16]:
```suggestion
```

File: unpin.js (hunk 5 above)
Comment (troggy) [inline unpin.js:20]:
```suggestion
        )
```

File: unpin.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+          });
+          //console.log(response.data.rows);
+            //handle response here
+        })
+        .catch(function (error) {
+            //handle error here
+        });
+}
+
+unPinExceptLast();
```
Comment (troggy) [inline unpin.js:44]:
```suggestion
const axios = require('axios');

const options = {
  headers: {
    'pinata_api_key': process.env.IPFS_DEPLOY_PINATA__API_KEY,
    'pinata_secret_api_key': process.env.IPFS_DEPLOY_PINATA__SECRET_API_KEY
  }
};

function deletePin(hashToUnpin){
  const url = `https://api.pinata.cloud/pinning/unpin/${hashToUnpin}`;
  return axios.delete(url, options);
}

function unPinExceptLast() {
  const url = `https://api.pinata.cloud/data/pinList?status=pinned&pageOffset=1`;
  return axios.get(url, options)
    .then(function (response) {
      response.data.rows.forEach(function(item){
        deletePin(item.ipfs_pin_hash)
      });
    });
}

unPinExceptLast();
```

**Comment by Clearwood** [comment]
> I've rerun a build without changes and unpin didn't work.
> 
> Checking pins with Postman.
> Before rerun there was just one pin for CID `QmQrNjf3xvnAcvR4gTSEKVKXyaRYV1SNnNujWynBpWw2uj` which is expected.
> After rerun there are two pins: new one and `QmQrNjf3xvnAcvR4gTSEKVKXyaRYV1SNnNujWynBpWw2uj`. I would expect to see either just the new one, or both but the old one with the `date_unpinned` set.
> 
> I suspect the reason for may be that `npm run build` terminates before the `deletePin` call happens because the `unpin` script is async. Or the script just not works by itself for some reason?

That is actually a feature not a bug :) The unpin happens before a new version is deployed so it does not delete the last pin but only pins leading up to that. Also I made the assumption that for each website a different account is used.

**Comment by Clearwood** [comment]
> cool work 
> 
> Should we remove S3 upload from deploy-prod.sh script already?

sure lets do that

File: scripts/deploy-prod.sh [hunk 8]
```diff
@@ -1,5 +1,5 @@
 #!/bin/sh
-
-aws s3 sync ./__sapper__/export s3://$PROD_S3_BUCKET/ --acl public-read
-aws configure set preview.cloudfront true
-aws cloudfront create-invalidation --distribution-id $PROD_CLOUDFRONT_DISTRIBUTION --paths "/*"
\ No newline at end of file
+./scripts/update-route53.sh
+#aws s3 sync ./__sapper__/export s3://$PROD_S3_BUCKET/ --acl public-read
+#aws configure set preview.cloudfront true
+#aws cloudfront create-invalidation --distribution-id $PROD_CLOUDFRONT_DISTRIBUTION --paths "/*"
```
Comment (troggy) [inline scripts/deploy-prod.sh:5]:
```suggestion
```

**Pull request merged**